src/hazelcore/semantics/OpSeq.re

```diff
@@ -70,3 +70,19 @@ let make_inconsistent =
     let set_skel = Skel.BinOp(InHole(TypeInconsistent, u), op, skel1, skel2);
     (OpSeq(set_skel, seq), u_gen);
   };
+
+let rec is_complete =
+        (
+          is_complete_operand: 'operand => bool,
+          opseq: t('operand, 'operator),
+        )
+        : bool => {
+  switch (opseq) {
+  | OpSeq(Placeholder(n) as _skel, seq) =>
+    is_complete_operand(seq |> Seq.nth_operand(n))
+  | OpSeq(BinOp(InHole(_), _, _, _), _) => false
+  | OpSeq(BinOp(NotInHole, _, skel1, skel2), seq) =>
+    is_complete(is_complete_operand, OpSeq(skel1, seq))
```

Comment:
you should define an auxiliary function is_complete_skel that takes a skel and a seq, rather than constructing intermediate OpSeqs (which are invalid as standalone OpSeqs and cause extra allocation)

Reply:
Why are standalone Opseqs invalid?

Reply:
oh in this case its the same seq so I guess its valid. okay, this is fine though it involves more allocations than doing it separately...

Reply:
Got it. I created a separate is_complete_skel function so that there are fewer allocations